deepmd/pt/model/model/pair_tab.py

```diff
@@ -0,0 +1,281 @@
+        )
+
+        # to calculate the atomic_energy, we need 3 tensors, i_type, j_type, rr
+        # i_type : (nframes, nloc), this is atype.
+        # j_type : (nframes, nloc, nnei)
+        j_type = extended_atype[
+            torch.arange(extended_atype.size(0))[:, None, None], masked_nlist
+        ]
+
+        # slice rr to get (nframes, nloc, nnei)
+        rr = torch.gather(pairwise_rr[:, :nloc, :], 2, masked_nlist)
+
+        raw_atomic_energy = self._pair_tabulated_inter(nlist, atype, j_type, rr)
+
+        atomic_energy = 0.5 * torch.sum(
+            torch.where(
+                nlist != -1, raw_atomic_energy, torch.zeros_like(raw_atomic_energy)
+            ),
+            dim=-1,
+        )
+
+        return {"energy": atomic_energy}
+
+    def _pair_tabulated_inter(
+        self,
+        nlist: torch.Tensor,
+        i_type: torch.Tensor,
+        j_type: torch.Tensor,
+        rr: torch.Tensor,
+    ) -> torch.Tensor:
+        """Pairwise tabulated energy.
+
+        Parameters
+        ----------
+        nlist : torch.Tensor
+            The unmasked neighbour list. (nframes, nloc)
+        i_type : torch.Tensor
+            The integer representation of atom type for all local atoms for all frames. (nframes, nloc)
+        j_type : torch.Tensor
+            The integer representation of atom type for all neighbour atoms of all local atoms for all frames. (nframes, nloc, nnei)
+        rr : torch.Tensor
+            The salar distance vector between two atoms. (nframes, nloc, nnei)
+
+        Returns
+        -------
+        torch.Tensor
+            The masked atomic energy for all local atoms for all frames. (nframes, nloc, nnei)
+
+        Raises
+        ------
+        Exception
+            If the distance is beyond the table.
+
+        Notes
+        -----
+        This function is used to calculate the pairwise energy between two atoms.
+        It uses a table containing cubic spline coefficients calculated in PairTab.
+        """
+        rmin = self.tab_info[0]
+        hh = self.tab_info[1]
+        hi = 1.0 / hh
+
+        self.nspline = int(self.tab_info[2] + 0.1)
+
+        uu = (rr - rmin) * hi  # this is broadcasted to (nframes,nloc,nnei)
+
+        # if nnei of atom 0 has -1 in the nlist, uu would be 0.
+        # this is to handel the nlist where the mask is set to 0, so that we don't raise exception for those atoms.
+        uu = torch.where(nlist != -1, uu, self.nspline + 1)
+
+        if torch.any(uu < 0):
+            raise Exception("coord go beyond table lower boundary")
+
+        idx = uu.to(torch.int)
+
+        uu -= idx
+
+        final_coef = self._extract_spline_coefficient(i_type, j_type, idx)
+
+        # here we need to do postprocess to overwrite coefficients when table values are not zero at rcut in linear extrapolation.
+        if self.tab.rmax < self.rcut:
+            post_mask = rr >= self.rcut
+            final_coef[post_mask] = torch.zeros(4)
+
```

Comment:
this differs from the tensorflow implementation, and may lead to inconsistent results. this is to handle a specific edge case, where rcut is slightly larger than the table upper boundary. for example, the table upper boundary is `0.2`, and the rcut is `0.22`, if the space between grids (hh) is `0.05`, in the current implementation, the table values do not decay to `0` at `0.22`, but at `0.25`. So for values between `0.2` and `0.22`, the values are calculated based on linear extrapolation in range 0.2~0.25, and the values at and beyond 0.22 are forced to `0`

Reply:
Note, this logic is not implemented in the Tensorflow version, need to add to the c++ operator